### PR TITLE
Map update develop merge fresh

### PR DESCRIPTION
**Pull request by greglemonmapbox**
**Related issue**

Runtime map updating

**Description of changes**

- Added EditorHelper class, which allows us to easily get class references in prop drawers
- Added MapboxDataProperty, which contains a settable HasChanged bool and a corresponding System.EventHandler that gets fired when HasChanged is set to true
- ModifierBase, LayerProperties and CoreVectorLayerProperties extend MapboxDataProperty
- PropertyDrawers for relevant AbstractMap UI classes use resulting bool return value of SerializedObject.ApplyModifiedProperties to set HasChanged to true
- Image section: Changing DataSource at runtime reloads image layer of map
- Terrain section: Changing DataSource, ElevationLayerType, and ExaggerationFactor reloads terrain and vector section of map

**Known issues**
- Vector section UI changes fire events; right now, VectorLayerVisualizer.UpdateVector listens for these events, but only displays debug message for now.
- Changing terrain UI causes tiles to be un/re-registered, which means VectorLayerVisualizer re-adds VectorLayerVisualizer.UpdateVector to the various layer properties PropertyHasChanged event, resulting in continually multiplying listener triggering, which can be seen in the resulting debug messages in the console window.

**QA checklists**

- [ ] Add relevant code comments. Every API class and method should have `<summary>` description as well as description of parameters.
- [ ] **Add tests for new/changed/updated classes and methods!!!**
- [ ] Check out conventions in [CONTRIBUTING.md](https://github.com/mapbox/mapbox-unity-sdk/blob/develop/CONTRIBUTING.md).
- [ ] Check out conventions in [CODING-STYLE.md](https://github.com/mapbox/mapbox-unity-sdk/blob/develop/CODING-STYLE.md)
- [ ] Update the [changelog](https://github.com/mapbox/mapbox-unity-sdk/blob/develop/documentation/docs/05-changelog.md)
- [ ] Update documentation.

**Reviewers**

Tag your reviewer(s). Choose wisely.

### CODE REVIEW

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/ElevatedTerrainStrategy.cs [hunk 1]
```diff
@@ -44,9 +44,15 @@ public override void RegisterTile(UnityTile tile)
 				tile.gameObject.layer = _elevationOptions.unityLayerOptions.layerId;
 			}
 
-			if (tile.MeshFilter.mesh.vertexCount == 0)
+			//_newVertexList.Count is the vertex count this strategy is expected to use
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/ElevatedTerrainStrategy.cs]: Is this comment relevant? the check is changed to a different thing.

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/LowPolyTerrainStrategy.cs [hunk 2]
```diff
@@ -47,9 +47,15 @@ public override void RegisterTile(UnityTile tile)
 				tile.gameObject.layer = _elevationOptions.unityLayerOptions.layerId;
 			}
 
-			if (tile.MeshFilter.mesh.vertexCount == 0)
+			//_newVertexList.Count is the vertex count this strategy is expected to use
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/LowPolyTerrainStrategy.cs]: Remove comment if not relevant.

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/VectorTileFactory.cs [hunk 3]
```diff
@@ -98,16 +114,33 @@ protected override void OnInitialized()
...
+					foreach (var item in layer.Value)
+					{
+						(item as VectorLayerVisualizer).SetProperties(null);
+						item.Initialize();
+					}
+				}
+			}
 		}
 
+		//OnColliderChange
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/VectorTileFactory.cs]: Remove?

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs [hunk 4]
```diff
@@ -143,55 +208,47 @@ public void SetProperties(VectorSubLayerProperties properties)
...
-						defaultGOModifierStack.Add(colorPaletteMod);
 					}
 					break;
 				default:
 					break;
 			}
 
-			_defaultStack.MeshModifiers.AddRange(defaultMeshModifierStack);
-			_defaultStack.GoModifiers.AddRange(defaultGOModifierStack);
+			//_defaultStack.MeshModifiers.AddRange(defaultMeshModifierStack);
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs]: Is this commented by mistake? Where are we adding default modifiers?

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/Modifiers/ModifierBase.cs [hunk 5]
```diff
@@ -4,8 +4,36 @@ namespace Mapbox.Unity.MeshGeneration.Modifiers
 	using System;
 	using Mapbox.Unity.MeshGeneration.Data;
 
+
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/Modifiers/ModifierBase.cs]: Move MapboxDataProperty to it's own file. Also, namespace for MapboxDataProperty should not be Mapbox.Unity.MeshGeneration.Modifiers

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/ElevatedTerrainStrategy.cs (hunk 1 above)
Comment (greglemonmapbox) [reply sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/ElevatedTerrainStrategy.cs]: @brnkhy

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/LowPolyTerrainStrategy.cs (hunk 2 above)
Comment (greglemonmapbox) [reply sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/LowPolyTerrainStrategy.cs]: @brnkhy

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs (hunk 4 above)
Comment (greglemonmapbox) [reply sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs]: @brnkhy

File: sdkproject/Assets/Mapbox/Unity/DataContainers/GeometryMaterialOptions.cs [hunk 6]
```diff
@@ -17,6 +17,8 @@ public override Type ModifierType
 			}
 		}
 
+		public int layerId;
```
[the remark this answers is not in the document]
Comment (greglemonmapbox) [reply sdkproject/Assets/Mapbox/Unity/DataContainers/GeometryMaterialOptions.cs]: @brnkhy

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainDataFetcher.cs [hunk 7]
```diff
@@ -1,20 +1,46 @@
...
+}
+
+public class ImageDataFetcherParameters : DataFetcherParameters
+{
+	public bool useRetina = true;
+}
+
+public class TerrainDataFetcherParameters : DataFetcherParameters
+{
+	public bool useRetina = true;
```
[the remark this answers is not in the document]
Comment (greglemonmapbox) [reply sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainDataFetcher.cs]: @brnkhy

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/ElevatedTerrainStrategy.cs (hunk 1 above)
Comment (brnkhy) [reply sdkproject/Assets/Mapbox/Unity/MeshGeneration/Factories/TerrainStrategies/ElevatedTerrainStrategy.cs]: no not anymore, that check will probably change soon as well

File: sdkproject/Assets/Mapbox/Unity/LayerProperties/ImageryLayerProperties.cs [hunk 8]
```diff
@@ -1,20 +1,56 @@
 ﻿namespace Mapbox.Unity.Map
 {
+	using UnityEngine;
+	using System;
+	using System.Collections.Generic;
+	using System.ComponentModel;
+	using System.Runtime.CompilerServices;
```
[the remark this answers is not in the document]
Comment (greglemonmapbox) [reply sdkproject/Assets/Mapbox/Unity/LayerProperties/ImageryLayerProperties.cs]: @brnkhy

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs (hunk 4 above)
Comment (brnkhy) [reply sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs]: @atripathi-mb it's adding default modifiers directly to the list. previously they were added to `defaultMeshModifierStack` and that list was added to the real mesh mod list using add range.

**Pull request merged**